tests/testsuite/warning_override.rs

```diff
@@ -0,0 +1,114 @@
+//! Tests for overriding warning behavior using `build.warnings` config option.
+
+use cargo_test_support::{cargo_test, project, Project};
+
+const ALLOW_CLEAN: &'static str = r#"[CHECKING] foo v0.0.1 ([ROOT]/foo)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+"#;
+
+const ALLOW_CACHED: &'static str = r#"[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+"#;
+
+const WARN: &'static str = r#"...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+"#;
+
+const DENY: &'static str = r#"...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[ERROR] warnings are denied by `build.warnings` configuration
+"#;
+
+fn make_project(main_src: &str) -> Project {
+    project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2021"
+            "#,
+        )
+        .file("src/main.rs", &format!("fn main() {{ {} }}", main_src))
+        .build()
+}
+
+#[cargo_test]
+fn rustc_caching_allow_first() {
+    let p = make_project("let x = 3;");
+    p.cargo("check")
+        .masquerade_as_nightly_cargo(&["warnings"])
+        .arg("-Zwarnings")
+        .arg("--config")
+        .arg("build.warnings='allow'")
+        .with_stderr_data(ALLOW_CLEAN)
+        .run();
+
+    p.cargo("check")
+        .masquerade_as_nightly_cargo(&["warnings"])
+        .arg("-Zwarnings")
+        .arg("--config")
+        .arg("build.warnings='deny'")
+        .with_stderr_data(DENY)
+        .with_status(101)
+        .run();
+}
+
+#[cargo_test]
+fn rustc_caching_deny_first() {
+    let p = make_project("let x = 3;");
+    p.cargo("check")
+        .masquerade_as_nightly_cargo(&["warnings"])
+        .arg("-Zwarnings")
+        .arg("--config")
+        .arg("build.warnings='deny'")
+        .with_stderr_data(DENY)
+        .with_status(101)
+        .run();
+
+    p.cargo("check")
+        .masquerade_as_nightly_cargo(&["warnings"])
+        .arg("-Zwarnings")
+        .arg("--config")
+        .arg("build.warnings='allow'")
+        .with_stderr_data(ALLOW_CACHED)
+        .run();
+}
+
+#[cargo_test]
+fn config() {
+    let p = make_project("let x = 3;");
+    p.cargo("check")
+        .masquerade_as_nightly_cargo(&["warnings"])
+        .arg("-Zwarnings")
+        .env("CARGO_BUILD_WARNINGS", "deny")
+        .with_stderr_data(DENY)
+        .with_status(101)
+        .run();
+
+    // CLI has precedence over config
```

Comment:
```suggestion
    // CLI has precedence over env
```
since they are both config